subprojects/core/src/main/java/org/gradle/caching/configuration/internal/DefaultBuildCacheConfiguration.java

```diff
@@ -96,25 +97,26 @@ public BuildCache getRemote() {
     }
 
     @Override
-    public void registerBuildCacheServiceFactory(BuildCacheServiceFactory<?> buildCacheServiceFactory) {
-        Preconditions.checkNotNull(buildCacheServiceFactory, "You cannot register a null build cache service factory.");
-        factories.put(buildCacheServiceFactory.getConfigurationType(), buildCacheServiceFactory);
+    public void registerBuildCacheService(Class<? extends BuildCache> configurationType, Class<? extends BuildCacheService> buildCacheServiceType) {
+        Preconditions.checkNotNull(configurationType, "configurationType cannot be null.");
+        Preconditions.checkNotNull(buildCacheServiceType, "buildCacheServiceType cannot be null.");
+        registeredTypes.put(configurationType, buildCacheServiceType);
     }
 
     @Override
-    public <T extends BuildCache> BuildCacheServiceFactory<T> getFactory(final Class<? extends T> buildCacheType) {
-        BuildCacheServiceFactory<?> factory = CollectionUtils.findFirst(factories.values(),
-            new Spec<BuildCacheServiceFactory<?>>() {
-                @Override
-                public boolean isSatisfiedBy(BuildCacheServiceFactory<?> factory) {
-                    return factory.getConfigurationType().isAssignableFrom(buildCacheType);
-                }
-            });
-        if (factory == null) {
-            throw new IllegalArgumentException(String.format("No build cache service factory for type '%s' could be found. Factories are known for %s.", buildCacheType.getSuperclass().getCanonicalName(), factories.keySet()));
+    public Class<? extends BuildCacheService> getBuildCacheServiceType(final BuildCache configuration) {
+        Map.Entry<Class<? extends BuildCache>, Class<? extends BuildCacheService>> matchingRegistration = CollectionUtils.findFirst(registeredTypes.entrySet(), new Spec<Map.Entry<Class<? extends BuildCache>, Class<? extends BuildCacheService>>>() {
+            @Override
+            public boolean isSatisfiedBy(Map.Entry<Class<? extends BuildCache>, Class<? extends BuildCacheService>> entry) {
+                Class<? extends BuildCache> configurationType = entry.getKey();
+                return configurationType.isInstance(configuration);
+            }
+        });
+
+        if (matchingRegistration == null) {
+            throw new IllegalArgumentException(String.format("No build cache service for configuration type '%s' could be found.", configuration.getClass().getSuperclass().getCanonicalName()));
         }
 
-        LOGGER.info("Loaded {} factory implementation {}", buildCacheType.getCanonicalName(), factory.getClass().getCanonicalName());
-        return Cast.uncheckedCast(factory);
+        return matchingRegistration.getValue();
```

Comment:
Is the functional approach worth pursuing here? I think this would be a lot easier to read with simple iteration.